SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
+	{
+		if(ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId = NewAuthVirtualWorkerId;
+		}
+	}));
+}
+
+void USpatialActorChannel::DropActorHierarchyMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId.Reset();
+		}
+	}));
+}
+
+void USpatialActorChannel::HandleActorLoadBalancing()
+{
+	// if the field is set it means we should try to migrate authority to another worker
+	if (MigrationVirtualWorkerId.IsSet())
+	{
+		AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+
+		// If we were locked in between replication frames, drop migration attempt.
+		if (NetDriver->LockingPolicy->IsLocked(Actor))
+		{
+			DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+		}
+		// Are we in a different frame from the one we determined we should try to migrate ?
+		else if (MigrationInitiatedFrame != NetDriver->ReplicationFrame)
+		{
+			// An invalid worker id means we encounter the first actor in the hierarchy, which will determine where all the other actors will migrate.
+			if (MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
 			{
-				OutTimestamp = Channel->AuthorityReceivedTimestamp;
+				const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*HierarchyRoot);
+				if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+				{
+					UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+					DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+				}
+				else if (NewAuthVirtualWorkerId == NetDriver->LoadBalanceStrategy->GetLocalVirtualWorkerId())
+				{
+					DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+				}
+				else
+				{
+					SetActorHierarchyMigrationDestination(NetDriver, NewAuthVirtualWorkerId, HierarchyRoot);
+				}
+			}
+
+			if (const VirtualWorkerId NewAuthVirtualWorkerId = MigrationVirtualWorkerId.GetValue())
+			{
+				Sender->SendAuthorityIntentUpdate(*Actor, MigrationVirtualWorkerId.GetValue());
+
+				// If we're setting a different authority intent, preemptively changed to ROLE_SimulatedProxy
+				Actor->Role = ROLE_SimulatedProxy;
+				Actor->RemoteRole = ROLE_Authority;
+
+				Actor->OnAuthorityLost();
+				MigrationVirtualWorkerId.Reset();
 			}
 		}
 	}
+	// Checking if we should retain authority.
+	else if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
+	{
+		AActor* TopmostOwner = SpatialGDK::GetTopmostOwner(Actor);
+		AActor* HierarchyRoot = TopmostOwner ? TopmostOwner : Actor;
```

Comment:
How come that the method to get `HierarchyRoot` is different here vs line 489? (`SpatialGDK::GetHierarchyRoot` vs this)